File: 2023/rust/src/bin/d05p2.rs
use std::fs;

#[derive(Debug)]
enum Maps {
    Seeds(Vec<(i64, i64)>),
    To(Vec<Mappings>),
}

#[derive(Debug)]
struct Mappings {
    dest_start: i64,
    src_start: i64,
    range_len: i64,
}

fn main() {
    println!("Day 5 - Part 2");

    let str = fs::read_to_string("../_input/day05.txt").unwrap();

    let total = process(&str);
    println!("Total: {}", total);
}

fn process(string: &str) -> i64 {
    let mut input = Vec::new();
    for line in string.lines() {
        input.push(line);
    }
    let input = input;

    // Load items
    let mut maps = Vec::new();
    let mut store = Vec::new();

    let mut i = 0;
    while i < input.len() {
        let l = input.get(i).unwrap();
        if l.is_empty() || i + 1 == input.len() {
            maps.push(convert_items(&store));
            store = Vec::new();
        } else {
            store.push(l);
        }

        i += 1;
    }

    // Do mappings
    let mut vals = match maps.remove(0) {
        Maps::Seeds(s) => s,
        _ => panic!("Seeds not at 0!"),
    };
    for m in maps {
        match m {
            Maps::To(s) => map_vals(&mut vals, s),
            _ => panic!("Not at 0!"),
        }
    }

    vals.iter().fold(i64::MAX, |acc, x| acc.min(x.0))
}

fn map_vals(vals: &mut Vec<(i64, i64)>, mappings: Vec<Mappings>) {
    let mut changes = vec![false; vals.len()];
    for m in mappings {
        let dest_start = m.dest_start;
        let dest_end = dest_start + m.range_len - 1;
        let src_start = m.src_start;
        let src_end = src_start + m.range_len - 1;
        let diff = dest_start - src_start;

        let mut i = 0;
        while i < vals.len() {
            if !changes[i] {
                let v1 = vals[i].0;
                let v2 = vals[i].1;

                if v1 >= src_start && v2 <= src_end {
                    // If the range is completly within, then just shift it.
                    vals[i].0 += diff;
                    vals[i].1 += diff;
                    changes[i] = true;
                } else if v1 >= src_start && v1 <= src_end && v2 > src_end {
                    // If the upper range is out of bounds
                    vals[i].0 += diff;
                    vals[i].1 = dest_end;

                    vals.push((src_end + 1, v2));

                    changes[i] = true;
                    changes.push(false);
                } else if v1 < src_start && v2 >= src_start && v2 <= src_end {
                    // If the lower range is out of bounds
                    vals[i].0 = dest_start;
                    vals[i].1 += diff;

                    vals.push((v1, src_start - 1));

                    changes[i] = true;
                    changes.push(false);
                }
            }

            i += 1;
        }
    }
}

fn convert_items(lines: &[&str]) -> Maps {
    let first = lines.first().unwrap();
    if first.starts_with("seeds:") {
        let mut seeds = first.split(' ').collect::<Vec<&str>>();
        seeds.remove(0);
        let seeds: Vec<i64> = seeds.iter().map(|i| str_to_i64(i)).collect();
        let seeds: Vec<(i64, i64)> = seeds
            .chunks_exact(2)
            .map(|i| (i[0], i[0] + i[1] - 1))
            .collect();
        Maps::Seeds(seeds)
    } else {
        Maps::To(make_mappings(&lines[1..lines.len()]))
    }
}

fn make_mappings(lines: &[&str]) -> Vec<Mappings> {
    lines
        .iter()
        .map(|i| {
            let vals: Vec<&str> = i.split(' ').collect();
            let vals: Vec<i64> = vals.iter().map(|i| str_to_i64(i)).collect();
            Mappings {
                dest_start: vals[0],
                src_start: vals[1],
                range_len: vals[2],
            }
        })
        .collect()
}

fn str_to_i64(str: &str) -> i64 {
    let mut acc = 0;
    for (i, c) in str.chars().enumerate() {
        acc += (10_i64.pow((str.len() - i - 1) as u32)) * c.to_digit(10).unwrap() as i64;
    }
    acc
}

#[cfg(test)]
mod test {
    use crate::process;

    #[test]
    fn test() {
        assert_eq!(
            process(
                r"seeds: 79 14 55 13

seed-to-soil map:
50 98 2
52 50 48

soil-to-fertilizer map:
0 15 37
37 52 2
39 0 15

fertilizer-to-water map:
49 53 8
0 11 42
42 0 7
57 7 4

water-to-light map:
88 18 7
18 25 70

light-to-temperature map:
45 77 23
81 45 19
68 64 13

temperature-to-humidity map:
0 69 1
1 0 69

humidity-to-location map:
60 56 37
56 93 4
"
            ),
            46
        );
    }
}

File: 2023/java/d05p2.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;

public class d05p2 {
    private static final int ZERO = 48;

    public static void main(String[] args) throws IOException {
        System.out.println("Day 5 - Part 2");

        BufferedReader reader = new BufferedReader(new FileReader("../_input/day05.txt"));

        long total = process(reader);
        reader.close();

        System.out.println("Total: " + total);
    }

    private static long process(BufferedReader reader) throws IOException {
        ArrayList<String> linesArr = new ArrayList<>();
        String line;
        while ((line = reader.readLine()) != null) {
            linesArr.add(line.trim());
        }
        String[] lines = linesArr.toArray(String[]::new);

        ArrayList<Seed> seeds = null;
        ArrayList<Mappings[]> mappings = new ArrayList<>();
        ArrayList<String> mapLines = new ArrayList<>();

        int i = 0;
        while (i < lines.length) {
            line = lines[i];
            if (line.isBlank() || i + 1 == lines.length) {
                if (mapLines.get(0).startsWith("seeds:")) seeds = convertSeeds(mapLines.get(0));
                else mappings.add(convertMap(mapLines.toArray(String[]::new)));
                mapLines = new ArrayList<>();
            } else mapLines.add(line);

            i++;
        }

        for (Mappings[] m : mappings) mapVals(seeds, m);

        long min = seeds.getFirst().start;
        for (var s : seeds) min = Math.min(min, s.start);
        return min;
    }

    private static void mapVals(ArrayList<Seed> vals, Mappings[] maps) {
        ArrayList<Boolean> changes = new ArrayList<>();
        for (int i = 0; i < vals.size(); i++) changes.add(false);

        for (var map : maps) {
            var destStart = map.destStart;
            var destEnd = destStart + map.rangeLen - 1;
            var srcStart = map.srcStart;
            var srcEnd = srcStart + map.rangeLen - 1;
            var diff = destStart - srcStart;

            for (int i = 0; i < vals.size(); i++) {
                if (!changes.get(i)) {
                    var val = vals.get(i);
                    var v1 = val.start;
                    var v2 = val.end;

                    if (v1 >= srcStart && v2 <= srcEnd) {
                        // If the range is completly within, then just shift it.
                        var seed = vals.get(i);
                        seed.start += diff;
                        seed.end += diff;
                        changes.set(i, true);
                    } else if (v1 >= srcStart && v1 <= srcEnd && v2 > srcEnd) {
                        // If the upper range is out of bounds
                        var seed = vals.get(i);
                        seed.start += diff;
                        seed.end = destEnd;

                        vals.add(new Seed(srcEnd + 1, v2));

                        changes.set(i, true);
                        changes.add(false);
                    } else if (v1 < srcStart && v2 >= srcStart && v2 <= srcEnd) {
                        // If the lower range is out of bounds
                        var seed = vals.get(i);
                        seed.start = destStart;
                        seed.end += diff;

                        vals.add(new Seed(v1, srcStart - 1));

                        changes.set(i, true);
                        changes.add(false);
                    }
                }
            }
        }
    }

    private static ArrayList<Seed> convertSeeds(String str) {
        String seedLine = str.split(":")[1].trim();
        long[] nums = getNums(seedLine);

        ArrayList<Seed> seeds = new ArrayList<>();
        for (int i = 0; i < nums.length; i += 2)
            seeds.add(new Seed(nums[i], nums[i] + nums[i + 1] - 1));
        return seeds;
    }

    private static Mappings[] convertMap(String[] items) {
        ArrayList<Mappings> maps = new ArrayList<>();
        for (int i = 1; i < items.length; i++) {
            long[] nums = getNums(items[i]);
            maps.add(new Mappings(nums[0], nums[1], nums[2]));
        }
        return maps.toArray(Mappings[]::new);
    }

    private static long[] getNums(String str) {
        String[] raws = str.split(" ");
        ArrayList<Long> builder = new ArrayList<>();
        for (var r : raws) {
            r = r.trim();
            if (!r.isEmpty()) builder.add(stringToLong(r));
        }
        return builder.stream().mapToLong(i -> i).toArray();
    }

    private static long stringToLong(String str) {
        long acc = 0;
        for (int i = 0; i < str.length(); i++) {
            long c = str.charAt(i);
            acc += (long) (Math.pow(10.0, str.length() - i - 1) * (c - ZERO));
        }
        return acc;
    }

    private static class Seed {
        public long start;
        public long end;

        public Seed(long start, long end) {
            this.start = start;
            this.end = end;
        }
    }

    private record Mappings(long destStart, long srcStart, long rangeLen) {}
}
